### ISSUE TITLE
Fine-Tuning LLama2 Loss is nan 

### ISSUE THREAD

**Issue by shailja-thakur** [facebookresearch/llama-recipes#146]
### System Info

I have been trying to fine-tune LLama-2-13B-chat using 3 A100*80GB each, however, the training log shows the loss is nan, any help will be appreciated

my training settings include the following,

model_name: str="PATH/to/LLAMA/7B"
    enable_fsdp: bool=False
    low_cpu_fsdp: bool=False
    run_validation: bool=True
    batch_size_training: int=4
    num_epochs: int=3
    num_workers_dataloader: int=1
    lr: float=1e-4
    weight_decay: float=0.0
    gamma: float= 0.85
    seed: int=42
    use_fp16: bool=False
    mixed_precision: bool=True
    val_batch_size: int=1
    dataset = "verilog_dataset"
    micro_batch_size: int=4
    peft_method: str = "lora" # None , llama_adapter, prefix
    use_peft: bool=False
    output_dir: str = "PATH/to/save/PEFT/model"
    freeze_layers: bool = False
    num_freeze_layers: int = 1
    quantization: bool = False
    one_gpu: bool = False
    save_model: bool = True
    dist_checkpoint_root_folder: str="PATH/to/save/FSDP/model" # will be used if using FSDP
    dist_checkpoint_folder: str="fine-tuned" # will be used if using FSDP
    save_optimizer: bool=False # will be used if using FSDP
    use_fast_kernels: bool = False # Enable using SDPA from PyTroch Accelerated Transformers, make use Flash Attention and Xformer memory-efficient kernels

and the command used is,

            torchrun \
                        --nnodes 1 \
                        --nproc_per_node 2  \
                        llama_finetuning.py \
                        --enable_fsdp \
                        --use_peft \
                        --peft_method lora \
                        --model_name llama-recipes/llama/llama-2-13b-chat-hf \
                        --pure_bf16 \
                        --lr 2e-4 \
                        --weight_decay 0.001 \
                        --output_dir llama-recipes/models \
                        --use_fast_kernels"

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### 🐛 Describe the bug

```
torchrun \
                        --nnodes 1 \
                        --nproc_per_node 2  \
                        llama_finetuning.py \
                        --enable_fsdp \
                        --use_peft \
                        --peft_method lora \
                        --model_name llama-recipes/llama/llama-2-13b-chat-hf \
                        --pure_bf16 \
                        --lr 2e-4 \
                        --weight_decay 0.001 \
                        --output_dir llama-recipes/models \
                        --use_fast_kernels"
                        ```

### Error logs

```Training Epoch0:  56%|█████▌    | 4648/8336 [1:26:08<1:09:13,  1.13s/it]
 step 4647 is completed and loss is nan
Training Epoch0:  56%|█████▌    | 4648/8336 [1:26:09<1:09:14,  1.13s/it]
Training Epoch0:  56%|█████▌    | 4649/8336 [1:26:09<1:08:55,  1.12s/it]
Training Epoch0:  56%|█████▌    | 4650/8336 [1:26:10<1:08:14,  1.11s/it]
 step 4649 is completed and loss is nan
Training Epoch0:  56%|█████▌    | 4650/8336 [1:26:11<1:08:14,  1.11s/it]
Training Epoch0:  56%|█████▌    | 4651/8336 [1:26:11<1:07:38,  1.10s/it]
Training Epoch0:  56%|█████▌    | 4652/8336 [1:26:12<1:07:59,  1.11s/it]
 step 4651 is completed and loss is nan
Training Epoch0:  56%|█████▌    | 4652/8336 [1:26:13<1:08:18,  1.11s/it]
Training Epoch0:  56%|█████▌    | 4653/8336 [1:26:13<1:07:38,  1.10s/it]
Training Epoch0:  56%|█████▌    | 4653/8336 [1:26:15<1:07:33,  1.10s/it]
Training Epoch0:  56%|█████▌    | 4654/8336 [1:26:14<1:07:18,  1.10s/it]
Training Epoch0:  56%|█████▌    | 4654/8336 [1:26:16<1:07:14,  1.10s/it]
 step 4654 is completed and loss is nan
Training Epoch0:  56%|█████▌    | 4655/8336 [1:26:15<1:07:42,  1.10s/it]
 step 4655 is completed and loss is nan
Training Epoch0:  56%|█████▌    | 4657/8336 [1:26:17<1:05:41,  1.07s/it]
 step 4656 is completed and loss is nan```


### Expected behavior

loss shoudl not be nan

**Comment by HamidShojanazeri** [comment]
@shailja-thakur do you have the same issue with BF16? its recommended to run with BF16 rather than fp16.

**Comment by shailja-thakur** [comment]
I am running the fine-tuning using bf16 only, but I guess my training dataset was not structured as per the models requirement, I am looking to fix that and will update you shortly if it worked or not

**Comment by shailja-thakur** [comment]
I have a question. what is the difference between "max_words" and "input_length", I have seen that alpaca dataset during pre-processing creates the "input_ids" of length max_words, however, the inputs would ideally use chunks of length input_length for processing, can you explain what is going on?

**Comment by shailja-thakur** [comment]
I was confusing max_words with # charas, I reduced the max_words limit to the max_words from the custom dataset, and it is working; however, may I ask what is the context length taken my the model if I have not passed the input context length explicitly?

**Comment by HumzaSami00** [comment]
@shailja-thakur  any luck ? I am facing same issue. Can you please share which dataset chat format did you used ?

**Comment by mitchelldehaven** [comment]
One thing to note here if you are using a custom dataset. You need to make that if you are cutting of portions of the input and labels based on length (i.e. the input length is too long) that the remaining labels you have are not all masked. If all labels you pass to the model are masked you will get NaN loss.

**Comment by wolframalpha** [comment]
> ### System Info
> I have been trying to fine-tune LLama-2-13B-chat using 3 A100*80GB each, however, the training log shows the loss is nan, any help will be appreciated
> 
> my training settings include the following,
> 
> model_name: str="PATH/to/LLAMA/7B" enable_fsdp: bool=False low_cpu_fsdp: bool=False run_validation: bool=True batch_size_training: int=4 num_epochs: int=3 num_workers_dataloader: int=1 lr: float=1e-4 weight_decay: float=0.0 gamma: float= 0.85 seed: int=42 use_fp16: bool=False mixed_precision: bool=True val_batch_size: int=1 dataset = "verilog_dataset" micro_batch_size: int=4 peft_method: str = "lora" # None , llama_adapter, prefix use_peft: bool=False output_dir: str = "PATH/to/save/PEFT/model" freeze_layers: bool = False num_freeze_layers: int = 1 quantization: bool = False one_gpu: bool = False save_model: bool = True dist_checkpoint_root_folder: str="PATH/to/save/FSDP/model" # will be used if using FSDP dist_checkpoint_folder: str="fine-tuned" # will be used if using FSDP save_optimizer: bool=False # will be used if using FSDP use_fast_kernels: bool = False # Enable using SDPA from PyTroch Accelerated Transformers, make use Flash Attention and Xformer memory-efficient kernels
> 
> and the command used is,
> 
> ```
>         torchrun \
>                     --nnodes 1 \
>                     --nproc_per_node 2  \
>                     llama_finetuning.py \
>                     --enable_fsdp \
>                     --use_peft \
>                     --peft_method lora \
>                     --model_name llama-recipes/llama/llama-2-13b-chat-hf \
>                     --pure_bf16 \
>                     --lr 2e-4 \
>                     --weight_decay 0.001 \
>                     --output_dir llama-recipes/models \
>                     --use_fast_kernels"
> ```
> 
> ### Information
> * [ ]  The official example scripts
> * [ ]  My own modified scripts
> 
> ### 🐛 Describe the bug
> ```
> torchrun \
>                         --nnodes 1 \
>                         --nproc_per_node 2  \
>                         llama_finetuning.py \
>                         --enable_fsdp \
>                         --use_peft \
>                         --peft_method lora \
>                         --model_name llama-recipes/llama/llama-2-13b-chat-hf \
>                         --pure_bf16 \
>                         --lr 2e-4 \
>                         --weight_decay 0.001 \
>                         --output_dir llama-recipes/models \
>                         --use_fast_kernels"
>                         ```
> 
> ### Error logs
> 
> ```Training Epoch0:  56%|█████▌    | 4648/8336 [1:26:08<1:09:13,  1.13s/it]
>  step 4647 is completed and loss is nan
> Training Epoch0:  56%|█████▌    | 4648/8336 [1:26:09<1:09:14,  1.13s/it]
> Training Epoch0:  56%|█████▌    | 4649/8336 [1:26:09<1:08:55,  1.12s/it]
> Training Epoch0:  56%|█████▌    | 4650/8336 [1:26:10<1:08:14,  1.11s/it]
>  step 4649 is completed and loss is nan
> Training Epoch0:  56%|█████▌    | 4650/8336 [1:26:11<1:08:14,  1.11s/it]
> Training Epoch0:  56%|█████▌    | 4651/8336 [1:26:11<1:07:38,  1.10s/it]
> Training Epoch0:  56%|█████▌    | 4652/8336 [1:26:12<1:07:59,  1.11s/it]
>  step 4651 is completed and loss is nan
> Training Epoch0:  56%|█████▌    | 4652/8336 [1:26:13<1:08:18,  1.11s/it]
> Training Epoch0:  56%|█████▌    | 4653/8336 [1:26:13<1:07:38,  1.10s/it]
> Training Epoch0:  56%|█████▌    | 4653/8336 [1:26:15<1:07:33,  1.10s/it]
> Training Epoch0:  56%|█████▌    | 4654/8336 [1:26:14<1:07:18,  1.10s/it]
> Training Epoch0:  56%|█████▌    | 4654/8336 [1:26:16<1:07:14,  1.10s/it]
>  step 4654 is completed and loss is nan
> Training Epoch0:  56%|█████▌    | 4655/8336 [1:26:15<1:07:42,  1.10s/it]
>  step 4655 is completed and loss is nan
> Training Epoch0:  56%|█████▌    | 4657/8336 [1:26:17<1:05:41,  1.07s/it]
>  step 4656 is completed and loss is nan```
> 
> 
> ### Expected behavior
> 
> loss shoudl not be nan 
> ```

While training a neural network, there can be multiple reasons for loss to become nan. 
One of the most common reasons, where I have experienced loss becoming nan is because we use high learning rate which cause model to diverge.
Reducing the learning rate to 5e-5 worked for me. Try it out.

**Comment by HamidShojanazeri** [comment]
@shailja-thakur, I believe this has been addressed not able to repro it. Please feel free to re-open if stilling running into issues.
